src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -251,351 +257,295 @@ internal NpgsqlTypeHandler GetByClrType(Type type)
-            NpgsqlTypeHandler elementHandler,
-            PostgresArrayType arrayPgType,
-            NpgsqlDbType? arrayNpgsqlDbType = null,
-            Type[]? elementClrTypes = null)
+        void ApplyUserMapping(PostgresType pgType, Type clrType, NpgsqlTypeHandler handler)
         {
-            Debug.Assert(Monitor.IsEntered(_writeLock));
-
-            NpgsqlTypeHandler arrayHandler = elementHandler.CreateArrayHandler(arrayPgType, _connector.Settings.ArrayNullabilityMode);
-
-            arrayHandler = GetOrBindHandler(arrayHandler, arrayPgType, arrayNpgsqlDbType);
-
-            // Note that array handlers aren't registered in ByClrType like base types, because they handle all
-            // dimension types and not just one CLR type (e.g. int[], int[,], int[,,]).
-            // So the by-type lookup is special and goes via _arrayHandlerByClrType, see this[Type type]
-            // TODO: register single-dimensional in _byType as a specific optimization? But avoid MakeArrayType for reflection-free mode?
-            if (elementClrTypes is not null)
-            {
-                foreach (var elementType in elementClrTypes)
-                {
-                    if (_arrayHandlerByClrType.TryGetValue(elementType, out var existingArrayHandler))
-                    {
-                        if (arrayHandler.GetType() != existingArrayHandler.GetType())
-                        {
-                            throw new Exception(
-                                $"Two array type handlers registered on same .NET type {elementType}: " +
-                                $"{existingArrayHandler.GetType().Name} and {arrayHandler.GetType().Name}");
-                        }
-                    }
-                    else
-                        _arrayHandlerByClrType[elementType] = arrayHandler;
-                }
-            }
+            _handlersByOID[pgType.OID] =
+                _handlersByDataTypeName[pgType.FullName] =
+                    _handlersByDataTypeName[pgType.Name] =
+                        _handlersByClrType[clrType] = handler;
 
-            return arrayHandler;
+            _userTypeMappings[pgType.OID] = new(npgsqlDbType: null, DbType.Object, pgType.Name, clrType);
         }
 
-        NpgsqlTypeHandler GetOrBindRangeHandler(NpgsqlTypeMapping subtypeMapping)
+        public override void AddTypeResolverFactory(ITypeHandlerResolverFactory resolverFactory)
         {
-            if (GetPostgresType(subtypeMapping).Range is not { } pgRangeType)
-                throw new ArgumentException($"No range type could be found in the database for subtype {subtypeMapping.PgTypeName}");
-
-            lock (_writeLock)
+            var sw = new SpinWait();
+            while (true)
             {
-                var subtypeHandler = GetOrBindBaseHandler(subtypeMapping);
-                var rangeHandler = subtypeHandler.CreateRangeHandler(pgRangeType);
+                var oldResolvers = _resolvers;
```

Comment:
`_resolvers` should be marked as `volatile`, otherwise jit can potentially move it outside of the cycle (it doesn't do it now, but...).